### PR TITLE
Handling empty reconciliation reports

### PR DESCRIPTION
**Pull request by apereverzin**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/DTSBPS-273

### Change description ###
ReconciliationSender makes decision if reconciliation report will be sent or not based on the value of skip-empty property.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailServiceTest.java [hunk 1]
```diff
@@ -416,7 +416,7 @@ void should_send_summary_report_mail_when_reconciliation_report_has_just_summary
         TargetStorageAccount account,
         ReconciliationReport reconciliationReport,
         SummaryReport summaryReport
-    ) throws IOException, SendEmailException {
+    ) throws Exception {
```
Comment (mdayican) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailServiceTest.java:419]: why?

File: src/main/resources/application.yaml [hunk 2]
```diff
@@ -157,6 +157,7 @@ reconciliation:
   report:
       mail-from: ${SMTP_USERNAME}
       mail-recipients: ${RECONCILIATION_REPORT_RECIPIENTS}
+      skip-empty: ${RECONCILIATION_REPORT_SKIP_EMPTY:true}
```
Comment (mdayican) [inline src/main/resources/application.yaml:160]: do we need configuration?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java [hunk 3]
```diff
@@ -49,28 +57,43 @@ public void sendReconciliationReport(
...
+
+            emailSender.sendMessageWithAttachments(
+                createTitle(account, summaryReport, detailedReport),
+                "",
+                mailFrom,
+                mailRecipients,
+                attachments
+            );
+        } else {
+            logger.info("");
```
Comment (mdayican) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java:84]: why empty?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java [hunk 4]
```diff
@@ -49,28 +57,43 @@ public void sendReconciliationReport(
...
-        File file = reconciliationCsvWriter.writeSummaryReconciliationToCsv(summaryReport);
-        attachments.put(
-            getReportAttachmentName(ATTACHMENT_SUMMARY_PREFIX, date),
-            file
-        );
-
-        if (detailedReport != null) {
-            File detailedReportFile = reconciliationCsvWriter
-                .writeDetailedReconciliationToCsv(detailedReport);
+        if (willReportsBeAttached(summaryReport, detailedReport)) {
```
Comment (mdayican) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java:60]: i think they just do not want to get empty file. if there is no discrepancy we do not need to add file?

File: src/test/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailServiceTest.java (hunk 1 above)
Comment (apereverzin) [reply src/test/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailServiceTest.java:419]: A bit not related to the task but we usually declare generic exceptions in test methods. Can revert for now but later I think it should be changed.

File: src/main/resources/application.yaml (hunk 2 above)
Comment (apereverzin) [reply src/main/resources/application.yaml:160]: Even if business does not need config right now I assume having this configurable will not be bad.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java (hunk 3 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java:84]: Oh! Will fix

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java (hunk 4 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java:60]: I decided they do not want to receive emails with empty report - emails is ket word. Will check with Sanchta.

File: src/main/resources/application.yaml (hunk 2 above)
Comment (mdayican) [reply src/main/resources/application.yaml:160]: I think we do not need. there is a config for email sending. without file there is no point to send email

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java (hunk 3 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java]: Added logs

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java [hunk 5]
```diff
@@ -49,44 +57,65 @@ public void sendReconciliationReport(
...
+                mailFrom,
+                mailRecipients,
+                attachments
+            );
+
+            logger.info(
+                "Email with reconciliation report has not been sent for {}, {} "
+                            + "because there are no discrepancies",
+                        account,
+                        date
```
Comment (mdayican) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationSender.java:102]: just above we are sending ?

**Review by mdayican — APPROVED** [review]

**Pull request merged**